opentracing/src/main/java/opentracing/SpanPropagator.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Encodes or Decodes a {@linkplain Span} in binary or text formats.
+ *
+ * <p>The toXXX methods are expected to serialize spans into a pair of values representing
+ * separately the span / span identity, and the trace attributes. This is done specifically
+ * for binary protocols that may represent tracing identity in a dedicated fixed-length slot in the
+ * binary message format, so that it can be inspected efficiently by the middleware / routing layers
+ * without parsing the whole message.
+ */
+public interface SpanPropagator {
+
+  /**
+   * Encodes the span into a binary representation of the span's identity and trace
+   * attributes.
+   */
+  byte[] toBinary(Span span);
```

Comment:
> For example Zipkin will provide OpenTracing instrumentation libraries to work with http and with thrift and what else.

This seems hardly practical given how many different RPC frameworks may be out there. Why should it be the duty of OepnTracing implementor and not of the implementor of the actual RPC framework? Even if we take http, there are hundreds of different server and client frameworks built on top of HTTP, all with their own APIs to prepare an outbound request and set its headers. The only commonality among them from OT point of view is that they deal with a text protocol, so OT provides a way to encode span into virtual "headers" as string key/value pairs. Then the actual framework decides how it can take that map of key/value pairs and translate it into the format of HTTP request it operates on.

Easier to show in the code: https://github.com/uber-common/opentracing-python-instrumentation/blob/master/opentracing_instrumentation/http_client.py#L59

This is an instrumentation layer that tries to be "generic" by relying on a certain abstraction of an http request. The actual http libraries all have different APIs for http request class, so adapters are provided to be able to reuse this "generic" wrapper around outbound request. This code is not an OpenTracing API, it's a user of that API.